### DISCUSSION TITLE
Shell running simultaneously both in WiFi and Serial

### DISCUSSION THREAD

**Discussion by maxbrito500** [dani007200964/Shellminator#15] [General]
Hello,

Progress is moving nicely, the WiFi version is really useful and Serial is also nice when debugging directly from the IDE.

It seems to be possible for usage of `commander.attachDebugChannel( &Serial );`, but would perhaps exist something similar as `commander.attachChannel( &Serial );` so that we can communicate with the same instance across multiple channels?

On the current library status it seems to only be possible by instantiating a second shell to be running.

Many thanks and wishing a good weekend!

**Comment by dani007200964** [comment]
Hi,

You can communicate with two or more channels, but you have to create a different shell object for each. It is because, the terminal has a lot of internal states, and I don't want to mess it up with different channels. ESP32 has plenty of memory to do that. You just create another Shellminator object for the Serial port and it can be used as well. You pass the same Commander object for both, and the instruction set will be the same.

Here is an example. I hope you can use something like this to achieve what you need 😃

```c

#define SERVER_PORT 23

// Create an instance of the server.
WiFiServer server( SERVER_PORT );

// Create a Shellminator object, and initialize it to use WiFiServer
Shellminator shellTCP( &server );

// Create a Shellminator object, and initialize it to use WiFiServer
Shellminator shellSerial( &Serial );

// We have to create an object from Commander class.
Commander commander;

// Commander API-tree
Commander::API_t API_tree[] = {
  // Add your commands like in the Commander-API Shellminator examples.
};


void setup(){

  // Init your peripherals like WiFi and Serial...

  // There is an option to attach a debug channel to Commander.
  // It can be handy to find any problems during the initialization
  // phase. In this example, we will use Serial for this.
  // It will be done only at startup, so it won't mess up the Serial terminal
  // at runtime.
  commander.attachDebugChannel( &Serial );

  commander.attachTree( API_tree );

  // Initialize Commander.
  commander.init();

  shellTCP.attachCommander( &commander );
  shellSerial.attachCommander( &commander );

  // initialize shell objects.
  shellTCP.begin( "brito" );
  shellSerial.begin( "brito" );

}

void loop(){
  shellTCP.update();
  shellSerial.update();
}

```

It will act like multiple terminal session for Linux. If you open two different terminals on a modern operating system, they work independently. But the instruction set is the same. This will do something similar. It is good because every channel is protected from the other one. It won't be good if you typed something on one session, and when you type something on the other one, it messes up the previous one. This way no crosstalk between different channels.

**Comment by maxbrito500** [comment]
Wow, really nice. Your rational makes good sense and the example is straightforward. I've started adopting that approach.

One question, some of the commands depend on the shell. Here is an example for "ll" (list files with details):


```cpp
// list the files inside the current folder
void func_ll(char *args, Stream *response ){
  File32 dir;
  File32 file;

  // open the current directory
  if (!dir.openCwd()){
    response -> println("Failed to open the directory");
    return;
  }

  while (file.openNext(&dir, O_RDONLY)) {
    char fileName[255];
    size_t len = file.getName8(fileName, sizeof(fileName));


    String time = getModifyDateTime(file);
    response -> print(time + " ");

    if(file.isDirectory()){
      // 34 == blue, 37 = white, 32 = green
      shell.setTerminalCharacterColor(response, 1, 34);
      response -> print(fileName);
      response -> print("/");
      shell.setTerminalCharacterColor(response, 0, 37);
      response -> print("\r\n");
    }else{
      response -> print(fileName);
      response -> print("\r\n");
    }
    file.close();
  }
  dir.close();
}
```

On this case how would you suggest to solve `shell.setTerminalCharacterColor(response, 0, 37);` situations that are specific to one shell instance in code that should be generic to all shells?

I'd assume one option is to repeat the same command for each shell instance but doesn't feel like proper programming since we might have even more shell instances in the future through other channels (e.g. throught bluetooth) so we'd then have to repeat three times each call to setTerminalCharacterColor().

What would be your thoughts on this topic?

**Reply by dani007200964** [reply]
This was a problem previously, and the solution is already implemented. If you check the `setTerminalCharacterColor` declarations in `Shellminator.hpp`, you can see that it has three versions. You will need this one:

```c
  /// Bring some color into your code.
  ///
  /// This function changes the color and style of the terminal application characters.
  /// This function can be used outside of a Shellminator object.
  /// @warning Please use the color and style enumeration table from this application as a parameter.
  /// @param stream_p Stream object to print the style code.
  /// @param style <a href="https://www.nayab.xyz/linux/escapecodes.html">VT100 compatible font styles</a>
  /// @param color <a href="https://www.nayab.xyz/linux/escapecodes.html">VT100 compatible color code</a>
  static void setTerminalCharacterColor( Stream *stream_p, uint8_t style, uint8_t color );
```

This is a static function inside the class, which means it is object independent. This way you can use this in any situation. It is designed for that specific problem, that you have. You just give a stream( channel ) as a parameter, and you can use it just like the other implementations.

Also, it is a good practice to use color and style enumerations instead of regular numbers:
```c
// Not as future-proof. Not recommended.
Shellminator::setTerminalCharacterColor( &Serial, 0, 32 );

// Futureproof, recommended, and easy to read.
Shellminator::setTerminalCharacterColor( &Serial, Shellminator::BOLD, Shellminator::RED );
```
You can find the style enumerations [here](https://dani007200964.github.io/Shellminator/html/class_shellminator.html#pub-types).

**Comment by maxbrito500** [comment]
Thanks, as usual that solved the topic for terminal character color. I will later put this on the syntax example so others can learn to make their code agnostic to channels.

One other command is the clear screen function:

```cpp
// similar to CLEAR function in Linux
void func_clear(char *args, Stream *response ){
  shell.clear();
  Shellminator::clear(response);
}
```

I've tried applying the same approach but there is no static version. Maybe for a next release?

**Reply by dani007200964** [reply]
Good point!
I add this to the ToDo for the next release. As far as I see, you have good ideas for future improvements. I made a [ToDo list](https://github.com/dani007200964/Shellminator/issues/14) for the new release. Feel free to use it. This way, I hopefully won't miss anything 😄

Of course, if you have any questions to discuss, feel free to ask them here 😃

**Comment by maxbrito500** [comment]
Thanks, reading that ToDo looks really promissing.

There is another detail, every command outputs an extra `\r\n` after outputting the final message. This makes it less intuitive because we need to calculate when it should be a `print` or `println`.

For example:
```cpp
// prints a line of text
void func_echo(char *args, Stream *response ){
  response -> print(args);
}
```

On this case we have to use `print` instead of `println `because otherwise it gets two lines of text when the command is completed and this doesn't look good. But when using `Serial.printf() `we know exactly what to expect. Or perhaps this is a behaviour specific to the WiFi shell where this command is most often running?

Anyways, sorry for the amount of topics. I'm really using shellminator intensively, it is so easy to work with. You've really took the heavy lifting from command line interfaces. This feedback are just minor things.

**Reply by dani007200964** [reply]
I'm happy, that you can use it! Feedback is always a good thing, in my opinion. I'm happy, that you share your ideas 😄

But I don't understand this one exactly. Can you describe it a bit more in detail? Or even better with some examples.

**Comment by maxbrito500** [comment]
Sure.

When I type `echo Hello World!` as seen on that code this is the result:
```
root:$ echo Hello World!
Hello World!
root:$ 
```

But if I would use println() instead of print() then the output would contain an extra line:
```
root:$ echo Hello World!
Hello World!

root:$ 
```

It seems that after executing a command a `\r\n` is added automatically. So I need to be careful and add only `print` commands on the last text that is outputted instead of `println` to avoid this double line situation. This is kind of different from the typical behaviour of Serial.

**Reply by dani007200964** [reply]
Yep, I understand, that this can be confusing. I liked this better, but my friends told me the same as you. Let me work on that. I also have many of other topics, like the PC simulator so that it might take some time.

**Comment by maxbrito500** [comment]
That's OK, just to report.

Another thing is the limitation of characters per command. I added a command to download an image from the internet but shellminator refused to accept the full length of the URL. I imagine this is for security to avoid overrunning the memory.

Is there a place where we can configure the max length of strings accepted?

**Reply by dani007200964** [reply]
It uses fixed-size buffers for security reasons. Of course, you can set the size of the buffer in the [Shellminator-Settings.hpp](https://github.com/dani007200964/Shellminator/blob/main/src/Shellminator-Settings.hpp). __SHELLMINATOR_BUFF_LEN__ is the length on characters for one line and __SHELLMINATOR_BUFF_DIM__ is the command history depth. Be careful, because it will use __SHELLMINATOR_BUFF_LEN__ times __SHELLMINATOR_BUFF_DIM__ bytes of memory for the buffers. Similarly, you have to configure Commander-API as well if you use it.

**Comment by maxbrito500** [comment]
Thanks again Daniel! I've learned where they are placed and will soon be contributing with docs to help others find this info in case of need. On my testing it is exactly as you mentioned in regards to higher memory usage, so had to reduce the command history in order to increase the permitted command size. To prevent these changes from being overwritten, I've just placed the #define on the source code file, and you had already included the #ifndef instruction on them.

Another thing, now both serials are working across WiFi and USB but the terminal console on either platformIO or ArduinoIDE does not seem to support color escape codes for text. When using them from the built-in terminal I'm seeing this on platformIO serial from terminal (pressing CTRL+ALT+S):
```
␛[1;32mwhoami␛[0;37m: Displays the current user
␛[1;32mwifiScan␛[0;37m: Search for available networks around.
␛[1;32mwifiStat␛[0;37m: Print WiFi information.

␛[1;32m␛[1;37m:␛[1;34m$␛[0;37m 
```
Keystrokes like backspace still work well. This just seems to affect the color escape codes. 

And this is what comes on the ArduinoIDE 2.x serial monitor:
```
[1;32mwhoami[0;37m: Displays the current user
[1;32mwifiScan[0;37m: Search for available networks around.
[1;32mwifiStat[0;37m: Print WiFi information.

[1;32m[1;37m:[1;34m$[0;37m 
```

PlatformIO has a separate Serial Monitor tool for connecting to serial that you can manually start, and from that tool I can see the output with the expected color formatting. However, most times developers want to compile and immediately interact with Serial that is available from the compile window rather than manually going to the proper tool.


Would you have any idea of how we can define the disabling of color codes for specific shells like Serial on this case?

WiFi remains the preferred work option for end-users connecting, USB serial is mostly for dev debugging purposes in either case so color isn't that relevant there.

**Reply by dani007200964** [reply]
There was a request for that in the past. It has been implemented with the __enableFormatting__ boolean variable. It is public, so it can be accessed from the outside. For example, if you want to disable formatting only on the Serial port:

```c

#define SERVER_PORT 23

// Create an instance of the server.
WiFiServer server( SERVER_PORT );

// Create a Shellminator object, and initialize it to use WiFiServer
Shellminator shellTCP( &server );

// Create a Shellminator object, and initialize it to use WiFiServer
Shellminator shellSerial( &Serial );

// We have to create an object from the Commander class.
Commander commander;

// Commander API-tree
Commander::API_t API_tree[] = {
  // Add your commands like in the Commander-API Shellminator examples.
};


void setup(){

  // Init your peripherals like WiFi and Serial...

  // There is an option to attach a debug channel to Commander.
  // It can be handy to find any problems during the initialization
  // phase. In this example, we will use Serial for this.
  // It will be done only at startup, so it won't mess up the Serial terminal
  // at runtime.
  commander.attachDebugChannel( &Serial );

  commander.attachTree( API_tree );

  // Initialize Commander.
  commander.init();

  shellTCP.attachCommander( &commander );
  shellSerial.attachCommander( &commander );

  // ---- This is the important part ----
  shellSerial.enableFormatting = false;

  // initialize shell objects.
  shellTCP.begin( "brito" );
  shellSerial.begin( "brito" );

}

void loop(){
  shellTCP.update();
  shellSerial.update();
}

```

I'm not sure that this will disable all VT-100 commands, but it will definitely disable color and font data.

**Comment by maxbrito500** [comment]
Thanks, that was fantastic as always.

I've tried it out immediately and resulted OK for the prompt, just not yet for the internal "help" command:
```
␛[1;32mwhoami␛[0;37m: Displays the current user
␛[1;32mwifiScan␛[0;37m: Search for available networks around.
␛[1;32mwifiStat␛[0;37m: Print WiFi information.

root:$ 
```

And on my own custom commands like LS that show a different color for folders:
```
if (file.isDirectory()) {
            Shellminator::setTerminalCharacterColor(response, Shellminator::BOLD, Shellminator::BLUE);
            response->print(fileName);
            response->print("/");
            Shellminator::setTerminalCharacterColor(response, Shellminator::REGULAR, Shellminator::WHITE);
            response->print(" ");
        } else {
            response->print(fileName);
            response->print(" ");
        }
```
Shellminator on that case will not read the boolean for the target shell output and still include the escape characters. Anyways, these things are just nitpicking. Reality is that shellminator is already at 110% of the needed functionality. Just discovered this morning that you had already added a path changing mechanism built inside. Outstanding.

**Reply by dani007200964** [reply]
These are important to me. Just keep going 😄 

Please report an issue about the enableFormatting bug. This way I wont forget to fix it.

**Comment by maxbrito500** [comment]
Done: https://github.com/dani007200964/Shellminator/issues/16
